protected/controllers/AdminSampleController.php

```diff
@@ -205,189 +189,188 @@ public function actionChoose() {
-            
             // From a model we will clone
-            $sampleAttr = new SampleAttribute();
-            $sampleAttr->sample_id = $model->id;
-            
-            foreach (explode('",', $model->attributesList) as $sAttr) {
-                $sAttr = str_replace('"', '', $sAttr);
-                $data = explode('=', $sAttr);
-                if (count($data) == 2) {
-                    
-                    // Get attribute, if not exist we create
-                    $attribute = Attribute::model()->findByAttributes(array('structured_comment_name' => trim($data[0])));
+            $sampleAttribute = new SampleAttribute();
+            $sampleAttribute->sample_id = $model->id;
+
+            foreach (explode('",', $model->attributesList) as $attributes) {
+                $attributes = str_replace('"', '', $attributes);
+                $attributeData = explode('=', $attributes);
+                if (count($attributeData) == 2) {
+                    // Get attribute model
+                    $attribute = Attribute::model()->findByAttributes(array('structured_comment_name' => trim($attributeData[0])));
                     if (!$attribute) {
-                        $attribute = new Attribute();
-                        $attribute->structured_comment_name = trim($data[0]);
-                        $attribute->attribute_name = str_replace('_', ' ', trim($data[0]));
-                        $attribute->definition = str_replace('_', ' ', trim($data[0]));
-                        $attribute->save();
+                        $model->addError('error', 'Attribute name for the input ' . $attributeData[0]. "=" . $attributeData[1] .' is not valid - please select a valid attribute name!');
                     }
-                    
                     // Let's save the new sample attribute
-                    $sampleAttribute = clone $sampleAttr;
-                    $sampleAttribute->value = trim($data[1]);
+                    $sampleAttribute = clone $sampleAttribute;
+                    $sampleAttribute->value = trim($attributeData[1]);
                     $sampleAttribute->attribute_id = $attribute->id;
-                    $sampleAttribute->save();
+                    if (!$sampleAttribute->save()) {
+                        foreach ($sampleAttribute->getErrors() as $errors) {
+                            foreach ($errors as $errorMessage) {
+                                $model->addError('error', $errorMessage);
+                            }
+                        }
+                    }
                 }
             }
         }
     }
+
+    /**
+     * @param $tax_id
+     * @param $model
+     * @return array|CActiveRecord|mixed|Species|null
+     */
+    private function getSpecies($tax_id, $model)
+    {
+        $species = Species::model()->findByAttributes(array('tax_id' => $tax_id));
+        if (!$species) {
+            $model->addError('error', 'Species id ' . $tax_id . ' is not found!');
+        }
```

Comment:
**thought**: This conditional statement seems like a validation rule for creating new Sample model objects such that the new sample must be of an existing species in GigaDB. In which case, this functionality should be in the Sample model class?

Reply:
Hi @pli888, @kencho51,

Validators' remit is scoped to when an object is about to be created or updated.
Once an object is persisted to the database, it's no longer the validators' jurisdiction.
Getters must assume that the data they try to retrieve has gone through validation already,
so I'd say there is no need to add any validation here.

The name of the function is confusing though because it's not a real getter:

Getters don't change the state of the database (that's the role of setters), they just "get", and this function is doing a `$model->save()` inside it on line 372 and calling `$this->updateSampleAttributes($model)` on line 369.

I think you should take those two lines that change the database state outside that function, and looking at what's left,
I actually think there is no point to it as its body can just be put as-is in the parent `try` block.

I think I understand what you were trying to do but the `actionCreate()` is not big, so there's not much benefit in trying to factor out its content.